### PR TITLE
Avoid implicit conversions to floats in dart:ui

### PR DESCRIPTION
**Pull request by dnfield**
Dart has a built in `double` type. We currently sometimes let Dart's FFI mechanisms automatically truncate this to float (like in the path API), and sometimes implicitly convert Dart doubles to C++ floats (aka SkScalar and impeller:Scalar). We also do implicit conversions when dealing with `Float64List` data, especially w.r.t. matrices. This causes problems when applications use very large double values which then get silently truncated to infinity, especially in matrices and values that end up being used in matrix concatenation.

This is solved by the following.

  - Enable `-Wimplicit-float-conversion` for dart:ui's C++ sources.
  - Add a bunch of `f`s to C++ literals.
  - Create a utility method for truncating doubles to floats in a well defined manner, avoiding truncation from a finite double value to an infinite float value.
  - Delete Tonic's converters from double to float, since these silently truncate to infinity before the engine C++ code even touches them.

Fixes https://github.com/flutter/flutter/issues/121299

### CODE REVIEW

File: lib/ui/floating_point.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+/// Converts a double to a float, truncating finite values that are larger than
+/// FLT_MAX or smaller than FLT_MIN to those values.
+static float SafeNarrow(double value) {
+  if (std::isinf(value) || std::isnan(value)) {
+    return static_cast<float>(value);
+  } else {
+    // Avoid truncation to inf/-inf.
+    return std::max(
+        std::numeric_limits<float>::lowest(),
+        std::min(std::numeric_limits<float>::max(), static_cast<float>(value)));
```
Comment (gaaclarke) [inline lib/ui/floating_point.h:22]: I think std::clamp can be used here for potentially faster code and easier to maintain.

Comment (dnfield) [reply lib/ui/floating_point.h:22]: Done

**Comment by dnfield** [comment]
In some offline conversation with @flar:

We may want to push the narrowing down til after DL has done its thing (and after the layer tree code has done its thing) to avoid premature precision loss. But that only really makes sense if we can get those types to use doubles as well, particularly around their matrix code (which today uses SkMatrix/SkM44, which are both float based).

**Comment by dnfield** [comment]
(I'm going to hold off on updating merge conflicts until I get more feedback)

**Comment by chinmaygarde** [comment]
Neither FLT_MAX or INFINITY is what the user expects. Using one of the other hardly matters. An assertion ought to be added earlier (closer to the user) where this sort of error may occur. I am not sure if this is an improvement.

**Comment by dnfield** [comment]
> Neither FLT_MAX or INFINITY is what the user expects. Using one or the other hardly matters. An assertion ought to be added earlier (closer to the user) where this sort of error may occur. I am not sure if this is an improvement.

Part of how I got to this point was trying to add that assertion. But what does it tell the user to do? Clamp to FLT_MAX? Something else entirely?

**Comment by chinmaygarde** [comment]
From the latest conversation in triage, I think a case can be made that this is less surprising than the current state of things. I'd really like the application author to be in control of the precision and how its clamped but I don't think that is feasible even in the medium to long term.

I think we should cleanup this patch and land this.

**Review by chinmaygarde — APPROVED** [review]

**Comment by tvolkert** [comment]
FYI, it looks like this may have introduced tiny differences in the AA pixel values -- check out the cluster around coordinates `x=732, y=154` in this [Skia Gold diff](https://flutter-gold.skia.org/diff?grouping=name%3Dmaterial.material.circular_progress_indicator.indeterminate%26source_type%3Dflutter&left=c49f16290fa88317f32f6e743ee99041&right=f7f846e916246a94f4a5d9b2924079de).

I approved the diff since it was not perceptible.

**Comment by tvolkert** [comment]
Note that if this did indeed cause those small differences, we're likely to see a bunch of Scuba failures when this rolls into g3 as well.

**Comment by dnfield** [comment]
Frob filed a bug against this PR that said it caused some minor scuba differences in google3.

**Pull request merged**